mu4/domain/notation/internal/msczmetareader.cpp

```diff
@@ -0,0 +1,356 @@
+#include "msczmetareader.h"
+
+#include <sstream>
+
+#include <QXmlStreamReader>
+#include <QFileInfo>
+#include <QBuffer>
+
+#include "log.h"
+#include "stringutils.h"
+#include "notationerrors.h"
+
+#include "thirdparty/qzip/qzipreader_p.h"
+
+using namespace mu;
+using namespace mu::domain::notation;
+
+RetVal<Meta> MsczMetaReader::readMeta(const io::path& filePath) const
+{
+    RetVal<Meta> meta;
+
+    QFileInfo fileInfo(io::pathToQString(filePath));
+    if (!fileInfo.exists()) {
+        LOGE() << "File not exists: " << filePath;
+        meta.ret = make_ret(Err::FileNotFound);
+        return meta;
+    }
+
+    auto compressed = [] (const QFileInfo& fileInfo) {
+        return fileInfo.suffix() == "mscz";
+    };
+
+    if (compressed(fileInfo)) {
+        meta = loadCompressedMsc(filePath);
+    } else {
+        QFile file(fileInfo.filePath());
+        if (!file.open(QFile::ReadOnly)) {
+            LOGE() << "Failed open file:" << filePath << file.errorString();
+            meta.ret = make_ret(Err::FileOpenError);
+            return meta;
+        }
+
+        QXmlStreamReader reader(file.readAll());
+        meta = doReadMeta(reader);
+    }
+
+    if (meta.val.fileName.isEmpty()) {
+        meta.val.fileName = fileInfo.baseName();
+    }
+
+    return meta;
+}
+
+RetVal<Meta> MsczMetaReader::loadCompressedMsc(const mu::io::path& filePath) const
+{
+    RetVal<Meta> meta;
+
+    QFile file(mu::io::pathToQString(filePath));
+    if (!file.open(QIODevice::ReadOnly)) {
+        LOGE() << "Failed open file: " << filePath;
+        meta.ret = make_ret(Err::FileOpenError);
+        return meta;
+    }
+
+    QByteArray data = file.readAll();
+
+    QBuffer buffer(&data);
+
+    MQZipReader zipReader(&buffer);
+
+    QString rootfile = readRootFile(&zipReader);
+    if (rootfile.isEmpty()) {
+        meta.ret = make_ret(Err::FileNoRootfile);
+        return meta;
+    }
+
+    QByteArray rootBuffer = zipReader.fileData(rootfile);
+    if (rootBuffer.isEmpty()) {
+        auto fil = zipReader.fileInfoList();
+        for (const MQZipReader::FileInfo& fi : fil) {
+            if (mu::strings::endsWith(fi.filePath.toStdString(), ".mscx")) {
+                rootBuffer = zipReader.fileData(fi.filePath);
+                break;
+            }
+        }
+    }
+
+    QXmlStreamReader xmlReader(rootBuffer);
+    meta = doReadMeta(xmlReader);
+
+    meta.val.thumbnail = loadThumbnail(&zipReader);
+
+    return meta;
+}
+
+void MsczMetaReader::doReadBox(QXmlStreamReader& xmlReader, RawMeta& meta) const
```

Comment:
The second argument (and function type also) is a bit confusing. Can doReadMeta fully prepare meta?